syntax = "proto3";

package NhdStats;

enum NHDStatus {
    NHD_STATUS_OK = 0;
    NHD_STATUS_ERR = 1;
}

message Empty {
    uint32 nothing = 1;
}

message NICInfoBrief {
    uint32 used_rx = 1;
    uint32 used_tx = 2;
}

message NodeInfo {
    string name = 1;
    uint32 free_cpus = 2;
    uint32 used_cpus = 3;
    uint32 free_gpus = 4;
    uint32 used_gpus = 5;
    uint32 free_hugepages = 6;
    uint32 used_hugepages = 7;
    uint32 total_pods = 8;
    repeated NICInfoBrief nic_info = 9;
}

message NodeStats {
    NHDStatus status = 1;
    repeated NodeInfo info = 2;
}

message SchedulerStats {
    NHDStatus status = 1;
    uint32 failed_schedule_count = 2;
}

message PodInfo {
    string name = 1;
    string node = 2;
    map<string, string> annotations = 3;
    string namespace = 4;
    repeated uint32 misc_cores = 5;
    repeated uint32 proc_cores = 6;
    repeated uint32 proc_helper_cores = 7;
    repeated uint32 gpus = 8;
    repeated string nic_macs = 9;
    uint32 hugepages = 10;
}

message PodStats {
    NHDStatus status = 1;
    repeated PodInfo info = 2;
}

message NodeReq {
    string name = 1;
}


message DetailedNodeStats {
    NHDStatus status = 1;
    string name = 2;
    repeated PodInfo podinfo = 3;
}

service NHDControl {
    rpc GetBasicNodeStats (Empty) returns (NodeStats) {}
    rpc GetSchedulerStats (Empty) returns (SchedulerStats) {}
    rpc GetPodStats(Empty) returns (PodStats) {}
    rpc GetDetailedNodeStats(NodeReq) returns (DetailedNodeStats) {}
}
